# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/frontend/BytecodeEmitter.cpp
# Commit: 7bd0784e7a6f
# Full Hash: 7bd0784e7a6f3702152bdca395b2305f16ca0479
# Author: Chris Fallin <cfallin@mozilla.com>
# Date: 2019-12-03 09:48:30
# Regressor Bug: 1598347
# File Overlap Count: 1
# Description:
#   Bug 1598347, part 2: pass "inner singleton" to NEWOBJECT group logic. r=djvj,iain
#   
#   This change passes through the "inner singleton" status of a particular
#   object literal to the group-assignment logic for JSOP_NEWOBJECT by
#   instead using a variant opcode, JSOP_NEWOBJECT_WITHGROUP.  "Inner
# ==============================================================================

diff -r 6da3c648995f -r 7bd0784e7a6f js/src/frontend/BytecodeEmitter.cpp
--- a/js/src/frontend/BytecodeEmitter.cpp	Wed Nov 27 22:48:40 2019 +0000
+++ b/js/src/frontend/BytecodeEmitter.cpp	Wed Nov 27 22:49:33 2019 +0000
@@ -8162,8 +8162,12 @@
     return false;
   }
 
-  bool success = singleton ? emitIndex32(JSOP_OBJECT, gcThingIndex)
-                           : emitIndex32(JSOP_NEWOBJECT, gcThingIndex);
+  bool isInnerSingleton = flags.contains(ObjLiteralFlag::IsInnerSingleton);
+
+  JSOp op = singleton
+                ? JSOP_OBJECT
+                : isInnerSingleton ? JSOP_NEWOBJECT_WITHGROUP : JSOP_NEWOBJECT;
+  bool success = emitIndex32(op, gcThingIndex);
   if (!success) {
     return false;
   }
@@ -8545,6 +8549,9 @@
     if (!useObjLiteralValues) {
       flags += ObjLiteralFlag::NoValues;
     }
+    if (isInner) {
+      flags += ObjLiteralFlag::IsInnerSingleton;
+    }
 
     // Use an ObjLiteral op. This will record ObjLiteral insns in the
     // objLiteralWriter's buffer and add a fixup to the list of ObjLiteral